cassandra/policies.py

```diff
@@ -316,6 +327,138 @@ def on_add(self, host):
+
+    def _rack(self, host):
+        return host.rack or self.local_rack
+
+    def _dc(self, host):
+        return host.datacenter or self.local_dc
+
+    def populate(self, cluster, hosts):
+        for (dc, rack), dc_hosts in groupby(hosts, lambda h: (self._dc(h), self._rack(h))):
+            self._live_hosts[(dc, rack)] = tuple(set(dc_hosts))
+
+        if not self.local_rack and not self.local_dc:
+            self._endpoints = [
+                endpoint
+                for endpoint in cluster.endpoints_resolved]
+
+        self._position = randint(0, len(hosts) - 1) if hosts else 0
+
+    def distance(self, host):
+        rack = self._rack(host)
+        dc = self._dc(host)
+        if rack == self.local_rack and dc == self.local_dc:
+            return HostDistance.LOCAL_RACK
+
+        if dc == self.local_dc:
+            return HostDistance.LOCAL
+
+        if not self.used_hosts_per_remote_dc:
+            return HostDistance.IGNORED
+        else:
+            dc_hosts = tuple(host for (datacenter, _), hosts in self._live_hosts.items() if datacenter == dc for host in hosts)
+            if not dc_hosts:
+                return HostDistance.IGNORED
+
+            if host in list(dc_hosts)[:self.used_hosts_per_remote_dc]:
+
+                return HostDistance.REMOTE
+            else:
+                return HostDistance.IGNORED
+
+    def make_query_plan(self, working_keyspace=None, query=None):
+        pos = self._position
+        self._position += 1
+
+        local_rack_live = self._live_hosts.get((self.local_dc, self.local_rack), ())
+        pos = (pos % len(local_rack_live)) if local_rack_live else 0
+        for host in islice(cycle(local_rack_live), pos, pos + len(local_rack_live)):
+            yield host
+
+        local_live = tuple(host for (datacenter, rack), hosts in self._live_hosts.items() if datacenter == self.local_dc and rack != self.local_rack for host in hosts)
+        pos = (pos % len(local_live)) if local_live else 0
+        for host in islice(cycle(local_live), pos, pos + len(local_live)):
+            yield host
+
+        # the dict can change, so get candidate DCs iterating over keys of a copy
+        other_dcs = [dc for (dc, _) in self._live_hosts.copy().keys() if dc != self.local_dc]
+        for dc in other_dcs:
+            remote_live = tuple(host for (datacenter, _), hosts in self._live_hosts.items() if datacenter == dc for host in hosts)
+            for host in remote_live[:self.used_hosts_per_remote_dc]:
+                yield host
+
+    def on_up(self, host):
+        # not worrying about threads because this will happen during
+        # control connection startup/refresh
+        if not self.local_rack and host.rack:
+            if host.endpoint in self._endpoints:
+                self.local_rack = host.rack
+                self.local_dc = host.datacenter
+                log.info("Using rack '%s' and datacenter '%s' for RackAwareRoundRobinPolicy (via host '%s'); "
```

Comment:
also, the text sounds like a warning, but the level is info.

I would rather we don't have warning like that, but errors when misconfigured

Reply:
I don't think it's a warning. From what I understand there are 2 valid ways to use this policy: you can explicitly pass DC and Rack to policy, or you can skip it in which case it will be taken from one of the contact points.
So this is info log about which route was taken.

Reply:
Exactly

Reply:
What is the point of using this class if not passing in a rack ? (It's the same with the DC one,  but I don't think we should change that)

Letting the driver randomly set rack or data center, would be confusing to users.

every time they connect they might get different results, i.e. latency and such, cause different DC or different RACK was selected.

or I don't fully understand what those fallbacks are for...

Reply:
I made it that way due to how it looks like in DC aware policy, but we can decide to take another road.

Reply:
I'm just trying to assume to role of a user here, and suggest we put guardrails to begin with (no need to wait for people to get bitten to put them in :))

but anyhow it's a bit of a guess work, until we'll have contact with some who actually need it

Reply:
To add some data points:
Rust Driver requires user to set dc and/or rack explicitly if they are to be preferred, there is no mechanism  to fetch it from contact points.
Java Driver 4.x seems to also require user to explicitly pass DC / Rack to enable DC / Rack awareness
Java Driver 3.x has similar mechanism to this PR

4.x and Rust don't have such mechanisms, so I think we should follow their lead